fem/fe.cpp

```diff
@@ -12604,7 +12604,7 @@ void ND_TriangleElement::CalcCurlShape(const IntegrationPoint &ip,
 const double ND_SegmentElement::tk[1] = { 1. };
 
 ND_SegmentElement::ND_SegmentElement(const int p, const int ob_type)
-   : VectorFiniteElement(1, Geometry::SEGMENT, p, p - 1,
+   : VectorFiniteElement(1, Geometry::SEGMENT, p, p,
```

Comment:
I don't think this is correct -- Nedelec `p` elements on edges have order `p-1`. For example, in the lowest order case, `p=1`, the tangential components are constants on the edges. This is also seen in the line:
```
      obasis1d(poly1d.GetBasis(p - 1, VerifyOpen(ob_type)))
```
below.

Reply:
I should have done a better job of documenting this.

The `ND_FECollection` of order `p` allocates the various `ND_*Element` classes by passing `p` to them as the order. It also returns `p` when `ND_FECollectxion::GetOrder()` is called. The `FiniteElementSpace::GetFE(int i)` method recently introduced an `MFEM_ASSERT` that checks that the `FE_Collection::GetOrder()` and `FiniteElement::GetOrder()` methods return the same integer.

Possible fixes were:
1) Disable this new check either completely or in this special case when the FiniteElement is an ND_SegmentElement.
2) Allocate an order `p+1` ND_SegmentElement in the order `p` ND_FECollection.
3) Increase the finite element order of `ND_SegmentElement` from `p-1` to `p`.
4) Perhaps something else...

Option 3 seemed to be the most consistent with the other `ND_*Element` classes which use the order perpendicular to the edge rather than along the edge.

Reply:
Thanks @mlstowell , this makes sense. 

Since this the only special case (right?) I would vote for actually modifying correctly the `MFEM_ASSERT` for the `ND_SegmentElement` (not removing it).

Reply:
I was thinking that the `GetOrder` methods were returning the order of convergence not the highest polynomial degree. A piecewise constant basis does converge at first order, doesn't it?

Reply:
`GetOrder` should be the polynomial degree of the space, ping: @jakubcerveny

Reply:
The assert was added in `var-order-space-dev` to make sure the element order in the space is the same thing as the FE order. I have always been confused about the order of the lower-order entities in ND, but perhaps we can modify the check to only apply to the actual elements (`FE->GetDim() == mesh->Dimension()`), not edges or faces (`FE->GetDim() > 0`)?

Reply:
We may end up with a related problem in `FiniteElementSpace::GetBE(int i)` for RT spaces. This method requests a `FiniteElement` from the `FE_Collection` with the same `order` as the neighboring interior element. However, for RT spaces the boundary elements are L2 elements which, by this definition, have lower order than the interior element.

Reply:
Perhaps there is a need for two separate integers in these classes. One to indicate that `FiniteElement` classes which share the same value of this parameter all have compatible basis functions. And a second integer to indicate the polynomial degree so that adequate integration rules can be chosen. These are two separate concerns after all.

Reply:
I propose we change `FiniteElementCollection::GetOrder()` to `FiniteElementCollection::GetConvergenceOrder()` (feel free to propose other names) and we add `FiniteElement::GetConvergenceOrder()` (same name as in `FiniteElementCollection` whatever we decide for that).  Then @jakubcerveny's variable order modifications can verify that the integers returned from these new methods agree.

With this change we can continue to use `FiniteElement::GetOrder()` in the manner that @v-dobrev and @tzanio intended i.e. as a guide to selecting integration rule orders.

Ping: @dylan-copeland and @psocratis .

Reply:
I prefer leaving the `GetOrder()` as is (corresponding to the polynomial degree), and if we were to introduce a new one, I would prefer a name that refers to the de Rham Complex (exact sequence) order. For example on the discrete level in 3D the sequence is `W^p (grad)->Q^p (curl)->V^p (div)->Y^p` , where the symbol `p` is not neccessarily the polynomial order (i.e., for Cartesian product elements this is a multi-index)

Reply:
@mlstowell or @jakubcerveny, where is this `MFEM_ASSERT`?

Reply:
@v-dobrev , it is in `FiniteElementSpace::GetFE(int i)`

Reply:
I have not followed everything here, but I just wanted to add that there was a (long) discussion about `GetOrder` and `GetFaceOrder` in https://github.com/mfem/mfem/pull/1817, see e.g. here https://github.com/mfem/mfem/pull/1817/commits/2a0b624fd1cdad101a4e3bcd721d93a7e9ddcab3.

Reply:
If I understand correctly, in master, `GetOrder` is the polynomial degree, and for constant-order spaces (same order on all elements, meaning no p-refinements), `FiniteElementSpace::GetFE` asserts that `GetOrder` is the same for the `FiniteElement` and the `FiniteElementCollection`. It seems simplest not to change these functions, if possible. What is needed to fix the issue Mark found is either logic to change or bypass the assertion in this special case, or to add a new type of order to be checked in the assertion. There have been suggestions for the new type of order to be order of convergence or the de Rham complex order, and those both sound fine to me. We could have some new order function in addition to `GetOrder`, without changing `GetOrder`. Initially, the new order function would only be used in this one assertion, but it could be useful in the future in other ways.

As an alternative, if we just want to fix this assertion that all elements have the same order as the collection in some sense, one possibility, without introducing any new functions or concepts, would be to check the number of DOFs. That is, replace `FE->GetOrder() == fec->GetOrder()` with
`FE->GetDof() == fec->GetNumDof(mesh->GetElementGeometry(i), fec->GetOrder())`
In the `ND_SegmentElement` case, this would be checking `p == p` although the results of `GetOrder` are different.

Reply:
Dylan, there's one issue with comparing the number of dofs: `fec->GetNumDof` and the method it uses, `DofForGeometry`, return the number of **interior** dofs. On the other hand, `FE->GetDof` returns all dofs including the ones on the boundary of the element.

One simple solution for the issue could be to replace https://github.com/mfem/mfem/blob/913361dd0db723d07e9c7492102f8b887acabc48/fem/fe_coll.cpp#L2520-L2523

with
```c++
 ND_FECollection::ND_FECollection(const int p, const int dim, 
                                  const int cb_type, const int ob_type) 
    : FiniteElementCollection(dim > 1 ? p : p-1) 
 { 
```
which sets the **correct order** based on `dim`. This will prevent us from using the collection `ND_FECollection(p,dim)`, `dim=2,3` with 1D meshes which has been valid so far (before we introduced variable orders). In other words, we'll be breaking backward compatibility but, in this case, that is probably acceptable.

I think it is important to remember (and we should document this in the base class `FiniteElementCollection`) that the number returned by `GetOrder` and the `p` parameter in the variable-order methods `Clone`, `GetFE`, `GetNumDof`, and `GetDofOrdering` are generally NOT the same as the `p` parameter used the constructors of derived classes. Also, the `p` parameter in the variable-order methods `GetFE`, `GetNumDof`, and `GetDofOrdering` is generally NOT the polynomial order for the element corresponding to the `geom` parameter.

Reply:
I pushed the above fix -- it should fix the unit test failure.

I also added some comments, as I was suggesting above -- @jakubcerveny you may want to take a look at these too.

@mlstowell, @dylan-copeland, @psocratis, or anyone else, let me know if this solution does not look right to you.